src/rules/prefer-destructured-store-props.ts

```diff
@@ -0,0 +1,105 @@
+import type * as ESTree from "estree"
+import type { TSESTree } from "@typescript-eslint/types"
+import type { AST } from "svelte-eslint-parser"
+import { createRule } from "../utils"
+import { getStringIfConstant } from "../utils/ast-utils"
+
+type NodeRecord = { property: string; node: TSESTree.MemberExpression }
+
+export default createRule("prefer-destructured-store-props", {
+  meta: {
+    docs: {
+      description:
+        "Destructure values from object stores for better change tracking & fewer redraws",
+      category: "Best Practices",
+      recommended: false,
+    },
+    hasSuggestions: true,
+    schema: [],
+    messages: {
+      useDestructuring: `Destructure {{property}} from store {{store}} for better change tracking & fewer redraws`,
+      fixUseDestructuring: `Using destructuring like $: ({ {{property}} } = {{store}}); will run faster`,
+    },
+    type: "suggestion",
+  },
+  create(context) {
+    let script: AST.SvelteScriptElement
+    const reports: NodeRecord[] = []
+
+    return {
+      [`SvelteScriptElement`](node: AST.SvelteScriptElement) {
+        script = node
+      },
+
+      // {$foo.bar}
+      // should be
+      // $: ({ bar } = $foo);
+      // {bar}
+      // Same with {$foo["bar"]}
+      [`MemberExpression[object.name=/^\\$/]`](
+        node: TSESTree.MemberExpression,
+      ) {
+        const property =
+          node.property.type === "Identifier"
+            ? node.property.name
+            : getStringIfConstant(node.property as ESTree.Expression)
+
+        if (!property) {
+          return
+        }
+
+        reports.push({ property, node })
```

Comment:
Could you add the following test case? I tested it and noticed a false positive.

```svelte
<script>
  $: foo2 = $store.foo;
</script>

{foo2}
```

Reply:
Also, could you add the following test case? I think member access in reactive statements should probably be ignored.

```svelte
<script>
  import store from "./store.js";
	
  let foo, bar
  $: {
    foo = $store.foo;
    bar = $store.bar;
  }
</script>

<div>
  foo: {foo + " " + Date.now()}
</div>
<div>
  bar: {bar + " " + Date.now()}
</div>
```

Ideally it would be nice to use reactive variables, but I think that could be checked with another new rule.

Reply:
> Could you add the following test case? I tested it and noticed a false positive.
> 
> ```svelte
> <script>
>   $: foo2 = $store.foo;
> </script>
> 
> {foo2}
> ```

I don't see the false positive here, can you elaborate?

https://svelte.dev/repl/1d043d248c0c4200b809ca84242301f1?version=3.49.0

Reply:
> Also, could you add the following test case? I think member access in reactive statements should probably be ignored.

I think they should still trigger, but they shouldn't have suggestions. Added tracking of imported & defined names to not suggest if there would be a collision.

Reply:
I don't think you'll see any difference in render times between using destructuring and assigning values via member access.

https://svelte.dev/repl/897da31522c3410596f0dbc4916e5014?version=3.49.0

In my opinion, these member accesses should not be reported as the purpose of this rule is to reduce render times.
In addition to this rule, I think it makes sense to add a new rule that corrects member access to the destructuring writing style. The purpose of the new rules is to enforce stylistic conventions.
In my opinion, it makes sense to have separate rules because they serve different purposes.

What do you think?

Reply:
I just noticed that the REPL you provided in #205 seems to have incorrect results due to typo.

```diff
-	<code>direct = obj.a</code><br />{obj.a + " " + Date.now()}	
+	<code>direct = obj.a</code><br />{direct + " " + Date.now()}	
```

You wrote the following, but when I fix the typo there is no difference in the timestamps.

> Notice how often the timestamp is changing on the "obj.a" or "direct = obj.a" versions vs the "({ a } = obj)" version? Each change in that timestamp indicates another redraw for the element.

Reply:
Looking at this result, in my opinion, prefer-reactive-destructuring may also be better split rules.

Reply:
The purpose of `prefer-reactive-destructuring` might be seen as enforcing stylistic conventions rather than reducing rendering times.

Reply:
If we split the rules, it doesn't seem necessary to add a stylistic rule.
The `prefer-destructuring` rule and the new `svelte/prefer-reactive-destructuring` rule seem to report them.

```svelte
<script>
  /* eslint svelte/prefer-reactive-destructuring: error */
  $: foo2 = $store.foo;
</script>
```

```svelte
<script>
/* eslint prefer-destructuring: error */

import store from "./store.js";
  
let foo, bar
$: {
  foo = $store.foo;
  bar = $store.bar;
}
</script>
```

[prefer-destructuring DEMO]

[prefer-destructuring DEMO]: https://ota-meshi.github.io/eslint-plugin-svelte/playground/#eJxFjkEOgkAMAL/SNJwIsvcVfUkvAsWsgZZ0y4nwd1dM9DiTOcyOg46MEbs8WFr9ThJq4DwncViNJ7bLyNltG3yzJM8IbKYGdSAhScuq5pBdjWEyXYCwDSe2r0x4JQEgmdlhUm2gfxhJFWH/+GLgBtW3LnDGpfjLAkUeJF347WGDts2cMe7H8Qb6vEHv